### ISSUE TITLE
Overriding a property via a statically declared field fails typecheck

### ISSUE THREAD

**Issue by WasabiFan** [microsoft/pyright#2943]
**Describe the bug**

I was attempting to use the suggestion here: https://stackoverflow.com/questions/2736255/abstract-attributes-in-python

Where a parent ABC requires that a child defines a field or property, and the child can choose which way it does it. Pyright reports an error when assigning a value in place of the property. (Of course, it would pass if it were made a property getter returning a constant.)

I'm not sure you will consider this a bug; it might be intentional.

**To Reproduce**

Type-check the following with default settings:

```python
from abc import ABC, abstractmethod

class Parent(ABC):
    @property
    @abstractmethod
    def some_statically_known_value(self) -> int:
        pass

class Child(Parent):
    some_statically_known_value = 5

if __name__ == "__main__":
    instance = Child()
    print(instance.some_statically_known_value)
```

Gives:

```
 "Literal[5]" is incompatible with "property"
```

**Expected behavior**

This is accepted.

Note that Pyright's behavior isn't _wrong_ as-is: we're overriding a property with a straight value, which are technically different types. But from a usage perspective, both are equivalent, assuming you don't consider the field containing a property object to be an inherent requirement for the API.

I don't know whether _you_ do or not. In my particular use-case, I'd like it if this passed type-check.

**Screenshots or Code**
See above

**VS Code extension or command-line**

Both

**Additional context**

**Comment by erictraut** [comment]
Yes, this behavior is intended. This has been discussed before. The semantics of a property are different from that of an attribute. They are not interchangeable from a type perspective.

**Comment by WasabiFan** [comment]
Ok, thanks. Is there an alternative pyright-accepted way to implement this pattern, where a child is required to declare a value but can do so using a field (both to avoid the boilerplate and to make it available on the class itself)?

**Comment by erictraut** [comment]
From a type safety standpoint, child classes should always be consistent with their parent class. If the parent exposes a property, the child should override with a property. If the parent exposes an instance variable, the child should use the instance variable from the parent.

**Comment by erictraut** [comment]
There is an exception made for [protocol classes](https://www.python.org/dev/peps/pep-0544/), which impose other limitations that make it type safe to intermix properties and attributes. That may provide the solution you're looking for. The writability needs to match in this case (e.g. if the protocol defines a property with a setter, the child can implement it with a read/write instance variable, but if the protocol defines a property without a setter, the child must declare the instance variable as `Final` so it's not writable).

**Comment by WasabiFan** [comment]
Thanks, that makes sense. These don't work great for my use-case but my use-case is also kind of weird. What I _really_ want is a static abstract property or field. Protocols _would_ work but I'd prefer the explicit inheritance if possible.

Thanks!
